### PR TITLE
Use JWT instead of machine token to be able recover workspaces

### PR DESCRIPTION
**Pull request by akorneta**
### What does this PR do?
Replaces machine tokens with JWT tokens.
Rewrites `MachineLoginFilter` agents/server in accordance to token changes.
Rewrite go agents auth mechanism to use JWT.
Adds signature keys manager, dao (generation of signature keys performs once on tomcat startup).
Verification flow for agents changed, so now they do not need to request ws-master to verify tokens.

Simplified interaction flow described in sequence diagram below:
machine auth 2

New tokens will be signed by `RSASSA-PKCS-v1_5` using `SHA-512` and for the initial stage, it won't be configurable. 
JWT will contain following data:
```json
#Header
{
  "alg": "RS512",
  "kind": "machine_token"
}
# Payload
{
  "wsid": "workspacekrh99xjenek3h571",
  "uid": "b07e3a58-ed50-4a6e-be17-fcf49ff8b242",
  "uname": "john",
  "jti": "06c73349-2242-45f8-a94c-722e081bb6fd"
}
# Signature
{}
```

### What issues does this PR fix or reference?
#7785 
#9028
#### Release Notes
n/a

### CODE REVIEW

**Comment by akorneta** [comment]
ci-build

**Comment by akorneta** [comment]
ci-build

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test-ocp

**Comment by akorneta** [comment]
ci-build

**Comment by akorneta** [comment]
ci-test-ocp

**Comment by akorneta** [comment]
ci-build

**Comment by akorneta** [comment]
ci-build

**Comment by akorneta** [comment]
ci-test-ocp

**Comment by akorneta** [comment]
ci-test

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/AbstractKeycloakFilter.java [hunk 1]
```diff
@@ -23,8 +28,19 @@
  */
 public abstract class AbstractKeycloakFilter implements Filter {
 
+  private static final String MACHINE_TOKEN_KIND = "machine_token";
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/AbstractKeycloakFilter.java]: Is it possible to use `org.eclipse.che.multiuser.machine.authentication.shared.Constants.MACHINE_TOKEN_KIND` value here?

File: multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/signature/SignatureAlgorithmEnvProvider.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.eclipse.che.commons.lang.Pair;
+
+/**
+ * Provides signature algorithm into machine environment.
+ *
+ * @author Anton Korneta
+ */
+public class SignatureAlgorithmEnvProvider implements EnvVarProvider {
+
+  private static final String SIGNATURE_ALGORITHM_ENV = "SIGNATURE_ALGORITHM";
```
Comment (sleshchenko) [inline multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/signature/SignatureAlgorithmEnvProvider.java]: I think it's better to name env variable in Che env var naming style, like `CHE_MACHINE_AUTH_SIGNATURE__ALGORITHM'.
Or something simpler but Che related `CHE_SIGNATURE_ALGORITHM`

File: multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/signature/SignatureKeyManager.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      final PrivateKey privateKey =
+          KeyFactory.getInstance(keyPair.getPrivateKey().getAlgorithm())
+              .generatePrivate(getKeySpec(keyPair.getPrivateKey()));
+      final PublicKey publicKey =
+          KeyFactory.getInstance(keyPair.getPublicKey().getAlgorithm())
+              .generatePublic(getKeySpec(keyPair.getPublicKey()));
+      return new KeyPair(publicKey, privateKey);
+    } catch (NoSuchAlgorithmException | InvalidKeySpecException ex) {
+      LOG.error("Failed to convert signature key pair to Java keys cause: {}", ex.getMessage());
+      return null;
```
Comment (sleshchenko) [inline multiuser/machine-auth/che-multiuser-machine-authentication/src/main/java/org/eclipse/che/multiuser/machine/authentication/server/signature/SignatureKeyManager.java]: Should it really return `null` here?

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
+--
```
Comment (sleshchenko) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql]: Shouldn't this file be placed in `6.0.3` folder?

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+-- Copyright (c) 2012-2016 Codenvy, S.A.
+-- All rights reserved. This program and the accompanying materials
+-- are made available under the terms of the Eclipse Public License v1.0
+-- which accompanies this distribution, and is available at
+-- http://www.eclipse.org/legal/epl-v10.html
+--
+-- Contributors:
+--   Codenvy, S.A. - initial API and implementation
```
Comment (sleshchenko) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql]: Please use new license with `Red Hat, Inc.`

**Review by sleshchenko — COMMENTED** [review]
In general looks good to me.
Please take a look my small comments and consider adding TCK test for `SignatureKeyDao` into  `postgresql-tck`.

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql (hunk 4 above)
Comment (riuvshin) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql]: :O

Comment (akorneta) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql]: I'll move scripts before merge

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test

File: multiuser/machine-auth/che-multiuser-machine-authentication-agent/src/main/java/org/eclipse/che/multiuser/machine/authentication/agent/MachineLoginFilter.java [hunk 6]
```diff
@@ -64,50 +69,71 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
+          response,
+          SC_UNAUTHORIZED,
+          format("Authentication on machine failed cause: '%s'", ex.getMessage()));
     }
   }
 
+  @VisibleForTesting
+  String getWorkspaceId() {
+    if (workspaceId == null) {
+      workspaceId = System.getenv().get("CHE_WORKSPACE_ID");
```
Comment (skabashnyuk) [inline multiuser/machine-auth/che-multiuser-machine-authentication-agent/src/main/java/org/eclipse/che/multiuser/machine/authentication/agent/MachineLoginFilter.java]: you can inject it as named variable ```@Named("env.CHE_WORKSPACE_ID") String workspaceId```

File: wsagent/wsagent-local/src/main/java/org/eclipse/che/MachinePublicKeyProvider.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * Provides instance of {@link PublicKey} created from environment.
+ *
+ * @author Anton Korneta
+ */
+public class MachinePublicKeyProvider implements Provider<PublicKey> {
+
+  @Override
+  public PublicKey get() {
+    try {
+      final String keyValue = System.getenv().get(SIGNATURE_PUBLIC_KEY_ENV);
```
Comment (garagatyi) [inline wsagent/wsagent-local/src/main/java/org/eclipse/che/MachinePublicKeyProvider.java]: Why not inject it using guice?

**Review by garagatyi — APPROVED** [review]
Looks good. Please take a look at my comments about different minor things.

File: multiuser/machine-auth/che-multiuser-machine-authentication-agent/src/main/java/org/eclipse/che/multiuser/machine/authentication/agent/MachineLoginFilter.java [hunk 8]
```diff
@@ -64,50 +71,63 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
-      ((HttpServletResponse) response)
-          .sendError(
-              HttpServletResponse.SC_UNAUTHORIZED,
-              "Authentication on machine failed, token " + machineToken + " is invalid");
-    } catch (ApiException apiEx) {
-      ((HttpServletResponse) response)
-          .sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, apiEx.getMessage());
-    } finally {
-      EnvironmentContext.reset();
+      final Jws<Claims> jws = Jwts.parser().setSigningKey(publicKey).parseClaimsJws(token);
```
[the remark this answers is not in the document]
Comment (akorneta) [reply multiuser/machine-auth/che-multiuser-machine-authentication-agent/src/main/java/org/eclipse/che/multiuser/machine/authentication/agent/MachineLoginFilter.java]: I'll try to simplify the logic or at least decouple it into separate methods

File: agents/go-agents/core/auth/auth.go [hunk 9]
```diff
@@ -109,33 +127,46 @@ func (handler cachingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
...
+		return fmt.Errorf("Authentication failed, due to kind: '%v' or workspace id: '%v' is wrong", kind, WorkspaceId)
 	}
 	return nil
 }
 
+// Supplies RSA public key for verification of given token,
+// when token 'alg' header is different or any problem occurs while retrieving key then error will be returned
+func rsaKeyFunc(token *jwt.Token) (interface{}, error) {
+	if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+		return nil, fmt.Errorf("token with unsupported signing method '%v' provided", token.Header["alg"])
```
[the remark this answers is not in the document]
Comment (akorneta) [reply agents/go-agents/core/auth/auth.go:161]: added [test](https://github.com/eclipse/che/pull/9016/files#diff-8072eb6c8c69bf1181e7343dcb65c1b9R96) for this case

File: multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/AbstractKeycloakFilter.java [hunk 10]
```diff
@@ -23,8 +30,20 @@
  */
 public abstract class AbstractKeycloakFilter implements Filter {
 
+  @Inject protected SignatureKeyManager signatureKeyManager;
+
   protected boolean shouldSkipAuthentication(HttpServletRequest request, String token) {
-    return token != null && token.startsWith("machine");
+    if (token == null) {
```
[the remark this answers is not in the document]
Comment (akorneta) [reply multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/AbstractKeycloakFilter.java]: added tests

File: wsagent/wsagent-local/src/main/java/org/eclipse/che/MachinePublicKeyProvider.java [hunk 11]
```diff
@@ -29,20 +29,24 @@
...
 public class MachinePublicKeyProvider implements Provider<PublicKey> {
 
+  private final String publicKey;
+  private final String algorithm;
+
+  @Inject
+  public MachinePublicKeyProvider(
+      @Named("env.CHE_MACHINE_AUTH_SIGNATURE__ALGORITHM") String alg,
+      @Named("env.CHE_MACHINE_AUTH_SIGNATURE__PUBLIC__KEY") String key) {
+    checkArgument(key != null && !key.isEmpty(), "Signature key for token validation is not found");
```
Comment (garagatyi) [inline wsagent/wsagent-local/src/main/java/org/eclipse/che/MachinePublicKeyProvider.java]: AFAIK env vars injected using this way don't support not set env vars. Have you tried it?

**Comment by garagatyi** [comment]
Thank you for the updates!

File: wsagent/wsagent-local/src/main/java/org/eclipse/che/MachinePublicKeyProvider.java (hunk 11 above)
Comment (akorneta) [reply wsagent/wsagent-local/src/main/java/org/eclipse/che/MachinePublicKeyProvider.java]: I'll check that, thx

**Comment by akorneta** [comment]
ci-build

**Comment by akorneta** [comment]
ci-test

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql (hunk 4 above)
Comment (sleshchenko) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.0.0/13__add_signature_key.sql]: I think it's good time to move it =)

File: wsmaster/integration-tests/postgresql-tck/pom.xml [hunk 12]
```diff
@@ -259,6 +270,7 @@
                             <includes>
                                 <include>**/tck/**</include>
                             </includes>
+                            <debugForkedProcess>true</debugForkedProcess>
```
Comment (sleshchenko) [inline wsmaster/integration-tests/postgresql-tck/pom.xml]: Can you elaborate why did you add it?

**Review by sleshchenko — COMMENTED** [review]
LGTM :+1:

**Review by sleshchenko — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test

**Comment by akorneta** [comment]
ci-test
